pkg/containerutil/container_network_manager.go

```diff
@@ -0,0 +1,555 @@
+	netType, err := nettype.Detect(netOpts.NetworkSlice)
+	if err != nil {
+		return nil, err
+	}
+
+	var manager NetworkOptionsManager
+	switch netType {
+	case nettype.None:
+		manager = &noneNetworkManager{globalOptions, netOpts}
+	case nettype.Host:
+		manager = &hostNetworkManager{globalOptions, netOpts}
+	case nettype.Container:
+		manager = &containerNetworkManager{globalOptions, netOpts}
+	case nettype.CNI:
+		manager = &cniNetworkManager{globalOptions, netOpts, nil}
+	default:
+		return nil, fmt.Errorf("unexpected container networking type: %q", netType)
+	}
+
+	return manager, nil
+}
+
+// No-op types.NetworkOptionsManager for network-less containers.
+type noneNetworkManager struct {
+	globalOptions types.GlobalCommandOptions
+	netOpts       types.NetworkOptions
+}
+
+// Returns a copy of the internal types.NetworkOptions.
+func (m *noneNetworkManager) GetNetworkOptions() types.NetworkOptions {
+	return m.netOpts
+}
+
+// Verifies that the internal network settings are correct.
+func (m *noneNetworkManager) VerifyNetworkOptions(_ context.Context) error {
+	// No options to verify if no network settings are provided.
+	return nil
+}
+
+// Performs setup actions required for the container with the given ID.
+func (m *noneNetworkManager) SetupNetworking(_ context.Context, _ string) error {
+	return nil
+}
+
+// Performs any required cleanup actions for the given container.
+// Should only be called to revert any setup steps performed in SetupNetworking.
+func (m *noneNetworkManager) CleanupNetworking(_ context.Context, _ containerd.Container) error {
+	return nil
+}
+
+// Returns the set of NetworkingOptions which should be set as labels on the container.
+func (m *noneNetworkManager) GetInternalNetworkingOptionLabels(_ context.Context) (types.NetworkOptions, error) {
+	return m.netOpts, nil
+}
+
+// Returns a slice of `oci.SpecOpts` and `containerd.NewContainerOpts` which represent
+// the network specs which need to be applied to the container with the given ID.
+func (m *noneNetworkManager) GetContainerNetworkingOpts(_ context.Context, _ string) ([]oci.SpecOpts, []containerd.NewContainerOpts, error) {
+	// No options to return if no network settings are provided.
+	return []oci.SpecOpts{}, []containerd.NewContainerOpts{}, nil
+}
+
+// types.NetworkOptionsManager implementation for container networking settings.
+type containerNetworkManager struct {
+	globalOptions types.GlobalCommandOptions
+	netOpts       types.NetworkOptions
+}
+
+// Returns a copy of the internal types.NetworkOptions.
+func (m *containerNetworkManager) GetNetworkOptions() types.NetworkOptions {
+	return m.netOpts
+}
+
+// Verifies that the internal network settings are correct.
+func (m *containerNetworkManager) VerifyNetworkOptions(_ context.Context) error {
+	// TODO: check host OS, not client-side OS.
+	if runtime.GOOS != "linux" {
+		return errors.New("container networking mode is currently only supported on Linux")
```

Comment:
I was surprised to see this.  Maybe I don't see how this is used?

Reply:
The initial PR's scope was abstracting the networking logic for all platforms and only adding minimal CNI networking support for Windows **for now** as that is [the only network setup](https://github.com/containerd/nerdctl/blob/main/hack/configure-windows-ci.ps1#L32) which is [tested on Windows](https://github.com/containerd/nerdctl/blob/main/.cirrus.yml#L38).